### PR TITLE
Fix "PVC Volume not detached if pod deleted via namespace deletion" issue

### LINKED ISSUE #29051: PVC Volume not detached if pod deleted via namespace deletion
**Problem:**
An attachable volume referenced by a pod via a PVC object (instead of directly) is not detached if the pod is deleted via namespace deletion (i.e. deleting the pod's namespace instead of the pod object directly).

**Repro steps:**

0. Create a new volume
  * `gcloud compute disks create --zone=us-central1-b  test-0b`
1. Create a new namespace.
  * `kubectl create ns testns`
    * `namespace "testns" created`
2. Create PV and PVC objects:
  * `kubectl create -f volumetest_pvc-pv.yaml`
    * `persistentvolume "pv-test-detach" created`
    * `persistentvolumeclaim "claim-test-detach" created`
3. Verify PV/PVC objects are in bound state
  * `kubectl get pv`
    * `pv-test-detach     50Gi       RWO           Bound     testns/claim-test-detach               2m`
4. Create pod object that references the volume via PVC
  * `kubectl create -f volumetest_pod_pvc.yaml`
    * `replicationcontroller "sleepypod" created`
5. Verify pod gets into running state without issue (i.e. volume is attached)
  * `kubectl get pods --namespace testns`
    * `sleepypod-0b0eq   1/1       Running   0          3m`
6. Delete the namespace
  * `kubectl delete ns testns`
    * `namespace "testns" deleted`
7. Wait for pod to terminate
  * `kubectl get pods --namespace testns`
8. Check if volume remains attached after a few minutes
  * Expected: Volume is detached.
  * Actual: Volume remains attached indefinitely.

<details>
`volumetest_pvc-pv.yaml` is:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name : pv-test-detach
spec:
  claimRef:
    name: claim-test-detach
    namespace: testns
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 50Gi
  persistentVolumeReclaimPolicy: Retain
  gcePersistentDisk:
    fsType: ext4
    pdName: test-0b
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: claim-test-detach
  namespace: testns
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 50Gi
```

And `volumetest_pod_pvc.yaml` is:
```
apiVersion: v1
kind: ReplicationController
metadata:
  name: sleepypod
  namespace: testns
spec:
  replicas: 1
  selector:
    name: sleepy
  template:
    metadata:
      labels:
        name: sleepy
    spec:
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: claim-test-detach
      containers:
      - name: sleepycontainer1
        image: saadali/sleepy:v0.2
        env:
        - name: "FOO"
          value: " "
        resources:
          limits:
            cpu: "0.002"
            memory: "4Mi"
        volumeMounts:
        - name: data
          mountPath: /data
          readOnly: false
```
</details>

**Workarounds:**

1. Delete the pod object directly before deleting the namespace.
  * In step 7 above do `kubectl delete -f volumetest_pod_pvc.yaml` instead
  * If the attachable volume is referenced by a pod using a PVC object, and the pod object is deleted directly (instead of the namespace first),  the volume is correctly detached.
2. Reference volume directly in pod (without a PVC object).
  * If the attachable volume is referenced by a pod directly (without a PVC object), namespace deletion results in the volume being correctly detached.

### PR DESCRIPTION
**Pull request by saad-ali**
Fixes #29051: "PVC Volume not detached if pod deleted via namespace deletion"

This PR:
- Fixes a bug in `desired_state_of_the_world_populator.go` to check the value of `exists` returned by the `podInformer` so that it can delete pods even if the delete event is missed (or fails).
- Reduces the desired state of the world populators sleep period from 5 min to 1 min (reducing the amount of time a volume would remain attached if a volume delete event is missed or fails).

### CODE REVIEW

**Comment by sijnc** [comment]
I believe this:

"Modify attach/detach controller to cache PVC/PV for each volume so even if these objects are deleted, pod deletion events can be processed without error, resulting in even faster detaches.
Add a test to verify this scenario (PVC volumes are detached on namespace deletion) and prevent future regressions."

...has been causing us a lot of pain using PDs on GKE. We have a script that launches our deployments, pv,pvc, etc. It creates pv/pvc first and then the deployment, however the script that tears down the deployment deletes the pv/pvc first before deleting the deployment. We never delete the namespace and we deploy into a non default namespace.

We randomly see PDs remaining attached for long periods after deleting, usually finally detaches in 5-15 minutes. We have a GKE engineer looking at the cluster so if you want any details about the findings let me know.

We have another issue (may not be related) where deployment times start at 5-10 seconds on a fresh cluster and then after a few days or a week the deployment times drop to 1-2 minutes and never get better. If we delete the cluster and re-create we get back to 5-10 seconds, but since it's GKE that's our only option to fix. Not sure if this is related but it seems like the PD issues don't happen until the cluster starts doing slow deployments so it might be worth looking into.

If this is the issue we've been seeing on GKE it's caused a lot of pain over the last couple months.

**Comment by sijnc** [comment]
@saad-ali: If you're interested the Google GKE support case is 09807892.

File: pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 		}
 		// retrieve the pod object from pod informer with the namespace key
 		informerPodObj, exists, err := dswp.podInformer.GetStore().GetByKey(dswPodKey)
-		if err != nil || informerPodObj == nil {
+		if err != nil || (exists && informerPodObj == nil) {
```
Comment (matchstick) [inline pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go]: @saad-ali DO we want to make this check more descriptive? Is one worth logging more than another?

Trying to understand:

if err != nil {
   glog.Errorf("podInformer GetBy key failed with an error....")
  continue
}

if (exists && informerPodObj == nil) {
    glog.Info("podInformer GetByKey found pod that is already deleted %q", dwsPodKey)
   continue
}

**Comment by saad-ali** [comment]
@matchstick PTAL

Will do the following in a follow-up (https://github.com/kubernetes/kubernetes/issues/29262):
  * Modify attach/detach controller to cache PVC/PV for each volume so even if these objects are deleted, pod deletion events can be processed without error, resulting in even faster detaches.
  * Add a test to verify this scenario (PVC volumes are detached on namespace deletion) and prevent future regressions.

**Comment by saad-ali** [comment]
Opened https://github.com/kubernetes/kubernetes/issues/29311 to track improvements to this code.

**Comment by matchstick** [comment]
@saad-ali Giving a LGTM for 1.3 but we really need something better for 1.4 as this code made us bump heads several times.

**Pull request merged**